README
NAME
    CGI::Application::Plugin::DBIx::Class - Access a DBIx::Class Schema from
    a CGI::Application

VERSION
    version 1.000100

SYNOPSIS
     use CGI::Application::Plugin::DBIx::Class ':all';

     sub cgiapp_init  {
         my $self = shift;

         $self->dbic_config({
            schema => 'MyApp::Schema',
            connect_info => {
               dsn => $data_source,
               user => $username,
               password => $password,
            },
         });
     }

     sub person {
        my $self   = shift;
        my $id     = $self->query->param('id');
        my $person = $self->schema->resultset('People')->find($id);
        # ...
     }

     sub people {
        my $self   = shift;
        my $people = $self->page_and_sort(
            $self->simple_search(
               $self->schema->resultset('People')
            )
        );
        # ...
     }

DESCRIPTION
    This module helps you to map various DBIx::Class features to CGI
    parameters. For the most part that means it will help you search, sort,
    and paginate with a minimum of effort and thought.

METHODS
  dbic_config
     $self->dbic_config({
       schema => MyApp::Schema->connect(@connection_data),
       connect_info => { ... },
     });

    You must run this method in setup or cgiapp_init to setup your schema.

    Valid arguments are:

     schema - Required, Name of DBIC Schema
     connect_info - Optional, these arguments are what are passed to connect, if
       this isn't passed and a C<dbh> method exists, that will be used
     ignored_params - Optional, Params to ignore when doing a simple search or sort,
        defaults to
     [qw{limit start sort dir _dc rm xaction}]

     page_size - Optional, amount of results per page, defaults to 25

  page_and_sort
     my $resultset = $self->schema->resultset('Foo');
     my $result = $self->page_and_sort($resultset);

    This is a helper method that will first sort (with "simple_sort") your
    data and then paginate it. Returns a resultset.

  paginate
     my $resultset = $self->schema->resultset('Foo');
     my $result = $self->paginate($resultset);

    Paginates the passed in resultset based on the following CGI parameters:

     start - first row to display
     limit - amount of rows per page

    Returns a resultset.

  schema
     my $schema = $self->schema;

    This is just a basic accessor method for your schema

  search
     my $resultset   = $self->schema->resultset('Foo');
     my $searched_rs = $self->search($resultset);

    Calls the controller_search method on the passed in resultset with all
    of the CGI parameters. I like to have this look something like the
    following:

     # Base search dispatcher, defined in MyApp::Schema::ResultSet
     sub _build_search {
        my $self           = shift;
        my $dispatch_table = shift;
        my $q              = shift;

        my %search = ();
        my %meta   = ();

        foreach ( keys %{$q} ) {
           if ( my $fn = $dispatch_table->{$_} and $q->{$_} ) {
              my ( $tmp_search, $tmp_meta ) = $fn->( $q->{$_} );
              %search = ( %search, %{$tmp_search} );
              %meta   = ( %meta,   %{$tmp_meta} );
           }
        }

        return $self->search(\%search, \%meta);
     }

     # search method in specific resultset
     sub controller_search {
        my $self   = shift;
        my $params = shift;
        return $self->_build_search({
              status => sub {
                 return { 'repair_order_status' => shift }, {};
              },
              part_id => sub {
                 return {
                    'lineitems.part_id' => { -like => q{%}.shift( @_ ).q{%} }
                 }, { join => 'lineitems' };
              },
              serial => sub {
                 return {
                    'lineitems.serial' => { -like => q{%}.shift( @_ ).q{%} }
                 }, { join => 'lineitems' };
              },
              id => sub {
                 return { 'id' => shift }, {};
              },
              customer_id => sub {
                 return { 'customer_id' => shift }, {};
              },
              repair_order_id => sub {
                 return {
                    'repair_order_id' => { -like => q{%}.shift( @_ ).q{%} }
                 }, {};
              },
           },$params
        );
     }

  sort
     my $resultset = $self->schema->resultset('Foo');
     my $result = $self->sort($resultset);

    Exactly the same as search, except calls controller_sort. Here is how I
    use it:

     # Base sort dispatcher, defined in MyApp::Schema::ResultSet
     sub _build_sort {
        my $self = shift;
        my $dispatch_table = shift;
        my $default = shift;
        my $q = shift;

        my %search = ();
        my %meta   = ();

        my $direction = $q->{dir};
        my $sort      = $q->{sort};

        if ( my $fn = $dispatch_table->{$sort} ) {
           my ( $tmp_search, $tmp_meta ) = $fn->( $direction );
           %search = ( %search, %{$tmp_search} );
           %meta   = ( %meta,   %{$tmp_meta} );
        } elsif ( $sort && $direction ) {
           my ( $tmp_search, $tmp_meta ) = $default->( $sort, $direction );
           %search = ( %search, %{$tmp_search} );
           %meta   = ( %meta,   %{$tmp_meta} );
        }

        return $self->search(\%search, \%meta);
     }

     # sort method in specific resultset
     sub controller_sort {
        my $self = shift;
        my $params = shift;
        return $self->_build_sort({
             first_name => sub {
                my $direction = shift;
                return {}, {
                   order_by => { "-$direction" => [qw{last_name first_name}] },
                };
             },
           }, sub {
          my $param = shift;
          my $direction = shift;
          return {}, {
             order_by => { "-$direction" => $param },
          };
           },$params
        );
     }

  simple_deletion
     $self->simple_deletion({ rs => 'Foo' });

    Deletes from the passed in resultset based on the following CGI
    parameter:

     to_delete - values of the ids of items to delete

    Valid arguments are:

     rs - resultset loaded into schema

    Note that this method uses the $rs->delete method, as opposed to
    $rs->delete_all

  simple_search
     my $searched_rs = $self->simple_search({ rs => 'Foo' });

    This method just searches on all of the CGI parameters that are not in
    the "ignored_params" with a like "%$value%". If there are multiple
    values it will make the search an "or" between the different values.

    Valid arguments are:

     rs - source loaded into schema

  simple_sort
     my $resultset = $self->schema->resultset('Foo');
     my $sorted_rs = $self->simple_sort($resultset);

    Sorts the passed in resultset based on the following CGI parameters:

     sort - field to sort by, defaults to primarky key
     dir  - direction to sort

SEE ALSO
    CGI::Application::Plugin::DBH

CREDITS
    Thanks to Micro Technology Services, Inc. for funding the initial
    development of this module.

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for CGI-Application-Plugin-DBIx-Class

1.000100  2011-05-20 15:40:42 CST6CDT
  - Merge 1.000000 and 0.100210 ... frew fail

1.000000  2011-05-20 15:35:33 CST6CDT
  - Remove dep on CGI::Application::Plugin::DBH
  - Fix terrible leak

0.100210  2010-01-21 19:54:05 CST6CDT
       - Get rid of some warnings and fix my silly labelling for newer versions
         of perl

0.100180  2010-01-18 22:24:54 CST6CDT
       - I am so bad at this.  Fix deps.

0.100160  2010-01-16 16:44:37 CST6CDT
       - Fix broken build (world writable bits, windows--)

0.093011 Wed Oct 28 13:05:00 CDT 2009
       - Yet another dependancy fix

0.093010 Wed Oct 28 00:00:00 CDT 2009
       - Allow for arrayrefs in simple_search

0.093000 Tue Oct 27 06:45:00 CDT 2009
       - Configurable page size
       - Clarify some pod

0.092900 Sat Oct 17 01:15:00 CDT 2009
       - Stop using BUILD_REQUIRES as it causes build failures most of the time
       - Basic pod fixes

0.092830 Sat Oct 10 01:47:00 CDT 2009
       - Add more missing deps
       - Fix width of example code in POD

0.092801 Tue Oct 06 22:38:00 CDT 2009
       - Add missing build requires directives

0.092800 Tue Oct 06 15:32:00 CDT 2009
       - Initial Release



t/lib/CAPExtTest/CGIApp.pm
package CAPExtTest::CGIApp;
our $VERSION = '0.092820';


use strict;
use warnings;

use parent 'CGI::Application';

use Readonly;
use CGI::Application::Plugin::DBH (qw/dbh_config dbh/);
use CAPExtTest::Schema;
use CGI::Application::Plugin::DBIx::Class ':all';
use CGI::Application::Plugin::ExtJS ':all';

Readonly our $DBFILE => 'test.db';
Readonly our $CONNECT_STR => "dbi:SQLite:dbname=$DBFILE";

sub cgiapp_init {
  my $self = shift;

  $self->dbh_config( $CONNECT_STR );

  $self->dbic_config({
     schema => 'CAPExtTest::Schema',
  });
}

sub setup {
    my $self = shift;

    $self->start_mode('test_mode');

    $self->run_modes(test_mode => 'test_mode' );
}

sub test_mode {
   return 1;
}

1;



t/lib/CAPExtTest/Schema.pm
package CAPExtTest::Schema;
our $VERSION = '0.092820';


use parent 'DBIx::Class::Schema';

__PACKAGE__->load_namespaces();

"I am not a number, I'm a free man!";


t/lib/CAPDBICTest/CGIApp.pm
package CAPDBICTest::CGIApp;
our $VERSION = '0.100210';
use strict;
use warnings;

use parent 'CGI::Application';

use CGI::Application::Plugin::DBH (qw/dbh_config dbh/);
use CAPDBICTest::Schema;
use CGI::Application::Plugin::DBIx::Class ':all';

our $DBFILE = 'test.db';
our $CONNECT_STR = "dbi:SQLite:dbname=$DBFILE";

sub cgiapp_init {
  my $self = shift;

  $self->dbh_config( $CONNECT_STR );

  $self->dbic_config({
     schema => 'CAPDBICTest::Schema',
  });
}

sub setup {
    my $self = shift;

    $self->start_mode('test_mode');

    $self->run_modes(test_mode => 'test_mode' );
}

sub test_mode {
  my $self = shift;
  return 1;
}

1;



Build.PL

use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
  module_name   => 'CGI::Application::Plugin::DBIx::Class',
  license       => 'perl',
  dist_abstract => "Access\ a\ DBIx\:\:Class\ Schema\ from\ a\ CGI\:\:Application",
  dist_name     => "CGI\-Application\-Plugin\-DBIx\-Class",
  dist_author   => [
"Arthur\ Axel\ \"fREW\"\ Schmidt\ \<frioux\+cpan\@gmail\.com\>",
  ],
  requires      => {
    "CGI::Application" => '0',
    "CGI::Application::Plugin::DBH" => '0',
    "parent" => '0',
    "DBIx::Class" => '0.08108',
  },
  build_requires      => {
    "Test::More" => '0.88',
    "UNIVERSAL" => '0',
    "SQL::Translator" => '0',
    "FindBin" => '0',
    "Test::Deep" => '0',
  },
  script_files => [ qw() ],
);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Access a DBIx::Class Schema from a CGI::Application',
  'AUTHOR' => 'Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>',
  'BUILD_REQUIRES' => {},
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.30'
  },
  'DISTNAME' => 'CGI-Application-Plugin-DBIx-Class',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'CGI::Application::Plugin::DBIx::Class',
  'PREREQ_PM' => {
    'CGI::Application' => '0',
    'DBIx::Class' => '0.08108',
    'FindBin' => '0',
    'SQL::Translator' => '0.11002',
    'Test::Deep' => '0',
    'Test::More' => '0.88',
    'UNIVERSAL' => '0',
    'parent' => '0'
  },
  'VERSION' => '1.000100',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/CGI/Application/Plugin/ExtJS.pm
package CGI::Application::Plugin::ExtJS;
our $VERSION = '0.092820';



# ABSTRACT: Convert paginated DBIx::Class::ResultSet's to JSON-style structures

use strict;
use warnings;

require Exporter;

use base qw(Exporter AutoLoader);

our @EXPORT_OK   = qw(&ext_paginate &ext_parcel);
our %EXPORT_TAGS = (all => [@EXPORT_OK]);

sub ext_paginate {
   my $self      = shift;
   my $resultset = shift;
   my $method    = shift || 'TO_JSON';
   return $self->ext_parcel(
      [map $_->$method, $resultset->all],
      $resultset->pager->total_entries,
   );
}

sub ext_parcel {
   my $self   = shift;
   my $values = shift;
   my $total  = shift || scalar @{$values};

   return {
      data  => $values,
      total => $total,
   };
}

1;





=pod

=head1 NAME

CGI::Application::Plugin::ExtJS - Convert paginated DBIx::Class::ResultSet's to JSON-style structures

=head1 VERSION

version 0.092820

=pod 

=head1 SYNOPSIS

  use CGI::Application::Plugin::ExtJS ':all';

  sub people {
     # ...
     my $json = $self->ext_paginate($paginated_rs);
     # ...
  }

  sub people_lite {
     # ...
     my $json = $self->ext_paginate($paginated_rs, sub {
        my $person = shift;
        return {
           first_name => $person->first_name,
           last_name => $person->last_name,
        }
     });
     # ...
  }

  # this will call the 'foo' method on each person and put the returned
  # value into the datastructure
  sub people_more_different {
     # ...
     my $json = $self->ext_paginate($paginated_rs, 'foo');
     # ...
  }

  sub programmers_do_it_by_hand {
     # ...
     my $data = [qw{foo bar baz}];
     my $total = 10;
     my $json = $self->ext_parcel($data, $total);
     # ...
  }

  # defaults total to amount of items passed in
  sub some_programmers_do_it_by_hand_partially {
     # ...
     my $data = [qw{foo bar baz}];
     my $json = $self->ext_parcel($data);
     # ...
  }

=head1 DESCRIPTION

This module is mostly for sending L<DBIx::Class> paginated data to ExtJS based javascript code.

=head1 METHODS

=head2 ext_paginate

  my $resultset = $self->schema->resultset('Foo');
  my $results   = $self->paginate($resultset);
  my $json      = $self->ext_paginate($resultset);
  my $json_str  = to_json($json);

=head3 Description

Returns a structure like the following from the ResultSet:

  {
     data  => [@results],
     total => $count_before_pagination
  }

=head3 Valid arguments are:

  rs      - paginated ResultSet to get the data from
  coderef - any valid scalar that can be called on the result object

=head2 ext_parcel

  my $items    = [qw{foo bar baz}];
  my $total    = 7;
  my $json     = $self->ext_parcel($data, $total);
  my $json_str = to_json($json);

=head3 Description

Returns a structure like the following:

  {
     data  => [@{$items}],
     total => $total || scalar @{$items}
  }

=head3 Valid arguments are:

  list  - a list of anything you want to be in the data structure
  total - whatever you want to say the total is.  Defaults to size of
          the list passed in.

=head1 SEE ALSO

L<CGI::Application::Plugin::DBIx::Class>.



=head1 AUTHOR

  Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut 



__END__


lib/CGI/Application/Plugin/DBIx/Class.pm
package CGI::Application::Plugin::DBIx::Class;
BEGIN {
  $CGI::Application::Plugin::DBIx::Class::VERSION = '1.000100';
}

# ABSTRACT: Access a DBIx::Class Schema from a CGI::Application

use strict;
use warnings;
use Carp 'croak';

require Exporter;

use parent qw(Exporter AutoLoader);

our @EXPORT_OK   = qw(&dbic_config &page_and_sort &schema &search &simple_search &simple_sort &sort &paginate &simple_deletion);
our %EXPORT_TAGS = (all => [@EXPORT_OK]);

sub dbic_config {
   $_[0]->{__dbic_connect_info} = $_[1]->{connect_info};
   if (!$_[0]->{__dbic_connect_info} && $_[0]->can('dbh') ) {
      my $dbh = $_[0]->dbh;
      $_[0]->{__dbic_connect_info} = sub { $dbh }
   }
   my $self = shift;
   my $config = shift;
   my $ignored_params = $config->{ignored_params} ||
      [qw{limit start sort dir _dc rm xaction}];

   $self->{__dbic_ignored_params} = { map { $_ => 1 } @{$ignored_params} };

   $self->{__dbic_schema_class} = $config->{schema} or croak 'you must pass a schema into dbic_config';

   $self->{__dbic_page_size} = $config->{page_size} || 25;
}

sub page_and_sort {
   my $self = shift;
   my $rs = shift;
   $rs = $self->simple_sort($rs);
   return $self->paginate($rs);
}

sub paginate {
   my $self     = shift;
   my $resultset = shift;
   # param names should be configurable
   my $rows = $self->query->param('limit') || $self->{__dbic_page_size};
   my $page =
      $self->query->param('start')
      ? ( $self->query->param('start') / $rows + 1 )
      : 1;

   return $resultset->search_rs( undef, {
      rows => $rows,
      page => $page
   });
}

sub schema {
   my $self = shift;
   if ( !$self->{schema} ) {
      $self->{schema} = $self->{__dbic_schema_class}->connect( $self->{__dbic_connect_info} );
   }
   return $self->{schema};
}

sub search {
   my $self = shift;
   my $rs_name = shift or croak 'required parameter rs_name for search was undefined';
   my %q       = $self->query->Vars;
   my $rs      = $self->schema->resultset($rs_name);
   # name of method should be configurable
   return $rs->controller_search(\%q);
}

sub sort {
   my $self = shift;
   my $rs_name = shift or croak 'required param to sort rs_name was undefined';
   my %q       = $self->query->Vars;
   my $rs      = $self->schema->resultset($rs_name);
   # name of method should be configurable
   return $rs->controller_sort(\%q);
}

sub simple_deletion {
   my $self = shift;
   my $params = shift;
   # param names should be configurable
   my @to_delete = $self->query->param('to_delete') or croak 'Required parameter (to_delete) undefined!';
   my $rs =
     $self->schema->resultset( $params->{rs} )
     ->search({
           id => { -in => \@to_delete }
     })->delete;
   return \@to_delete;
}

sub simple_search {
   my $self = shift;
   my $params = shift;
   my $rs  = $params->{rs};
   my %skips  = %{$self->{__dbic_ignored_params}};
   my $searches = {};
   foreach ( keys %{ $self->query->Vars } ) {
      # we use '' here because it's conceivable that someone
      # would want to search with a 0, whereas '' is always
      # going to imply null for a user
      if ( $self->query->param($_) ne '' and not $skips{$_} ) {
         # should be configurable
         $searches->{$_} = { like => [map "%$_%", $self->query->param($_)]};
      }
   }

   my $rs_full = $self->schema()->resultset($rs)->search($searches);

   return $self->page_and_sort($rs_full);
}

sub simple_sort {
   my $self = shift;
   # param names should be configurable
   my $rs = shift or croak 'required parameter rs for simple_sort not defined';
   my %order_by = ( order_by => [ $rs->result_source->primary_columns ] );
   if ( $self->query->param('sort') ) {
      %order_by = (
         order_by => {
            q{-}.$self->query->param('dir') => $self->query->param('sort')
         }
      );
   }
   return $rs->search(undef, { %order_by });
}

1;




=pod

=head1 NAME

CGI::Application::Plugin::DBIx::Class - Access a DBIx::Class Schema from a CGI::Application

=head1 VERSION

version 1.000100

=head1 SYNOPSIS

 use CGI::Application::Plugin::DBIx::Class ':all';

 sub cgiapp_init  {
     my $self = shift;

     $self->dbic_config({
        schema => 'MyApp::Schema',
        connect_info => {
           dsn => $data_source,
           user => $username,
           password => $password,
        },
     });
 }

 sub person {
    my $self   = shift;
    my $id     = $self->query->param('id');
    my $person = $self->schema->resultset('People')->find($id);
    # ...
 }

 sub people {
    my $self   = shift;
    my $people = $self->page_and_sort(
        $self->simple_search(
           $self->schema->resultset('People')
        )
    );
    # ...
 }

=head1 DESCRIPTION

This module helps you to map various L<DBIx::Class> features to CGI parameters.
For the most part that means it will help you search, sort, and paginate with a
minimum of effort and thought.

=head1 METHODS

=head2 dbic_config

 $self->dbic_config({
   schema => MyApp::Schema->connect(@connection_data),
   connect_info => { ... },
 });

You must run this method in setup or cgiapp_init to setup your schema.

Valid arguments are:

 schema - Required, Name of DBIC Schema
 connect_info - Optional, these arguments are what are passed to connect, if
   this isn't passed and a C<dbh> method exists, that will be used
 ignored_params - Optional, Params to ignore when doing a simple search or sort,
    defaults to
 [qw{limit start sort dir _dc rm xaction}]

 page_size - Optional, amount of results per page, defaults to 25

=head2 page_and_sort

 my $resultset = $self->schema->resultset('Foo');
 my $result = $self->page_and_sort($resultset);

This is a helper method that will first sort (with C<simple_sort>) your data and
then paginate it.  Returns a resultset.

=head2 paginate

 my $resultset = $self->schema->resultset('Foo');
 my $result = $self->paginate($resultset);

Paginates the passed in resultset based on the following CGI parameters:

 start - first row to display
 limit - amount of rows per page

Returns a resultset.

=head2 schema

 my $schema = $self->schema;

This is just a basic accessor method for your schema

=head2 search

 my $resultset   = $self->schema->resultset('Foo');
 my $searched_rs = $self->search($resultset);

Calls the controller_search method on the passed in resultset with all of the
CGI parameters.  I like to have this look something like the following:

 # Base search dispatcher, defined in MyApp::Schema::ResultSet
 sub _build_search {
    my $self           = shift;
    my $dispatch_table = shift;
    my $q              = shift;

    my %search = ();
    my %meta   = ();

    foreach ( keys %{$q} ) {
       if ( my $fn = $dispatch_table->{$_} and $q->{$_} ) {
          my ( $tmp_search, $tmp_meta ) = $fn->( $q->{$_} );
          %search = ( %search, %{$tmp_search} );
          %meta   = ( %meta,   %{$tmp_meta} );
       }
    }

    return $self->search(\%search, \%meta);
 }

 # search method in specific resultset
 sub controller_search {
    my $self   = shift;
    my $params = shift;
    return $self->_build_search({
          status => sub {
             return { 'repair_order_status' => shift }, {};
          },
          part_id => sub {
             return {
                'lineitems.part_id' => { -like => q{%}.shift( @_ ).q{%} }
             }, { join => 'lineitems' };
          },
          serial => sub {
             return {
                'lineitems.serial' => { -like => q{%}.shift( @_ ).q{%} }
             }, { join => 'lineitems' };
          },
          id => sub {
             return { 'id' => shift }, {};
          },
          customer_id => sub {
             return { 'customer_id' => shift }, {};
          },
          repair_order_id => sub {
             return {
                'repair_order_id' => { -like => q{%}.shift( @_ ).q{%} }
             }, {};
          },
       },$params
    );
 }

=head2 sort

 my $resultset = $self->schema->resultset('Foo');
 my $result = $self->sort($resultset);

Exactly the same as search, except calls controller_sort.  Here is how I use it:

 # Base sort dispatcher, defined in MyApp::Schema::ResultSet
 sub _build_sort {
    my $self = shift;
    my $dispatch_table = shift;
    my $default = shift;
    my $q = shift;

    my %search = ();
    my %meta   = ();

    my $direction = $q->{dir};
    my $sort      = $q->{sort};

    if ( my $fn = $dispatch_table->{$sort} ) {
       my ( $tmp_search, $tmp_meta ) = $fn->( $direction );
       %search = ( %search, %{$tmp_search} );
       %meta   = ( %meta,   %{$tmp_meta} );
    } elsif ( $sort && $direction ) {
       my ( $tmp_search, $tmp_meta ) = $default->( $sort, $direction );
       %search = ( %search, %{$tmp_search} );
       %meta   = ( %meta,   %{$tmp_meta} );
    }

    return $self->search(\%search, \%meta);
 }

 # sort method in specific resultset
 sub controller_sort {
    my $self = shift;
    my $params = shift;
    return $self->_build_sort({
         first_name => sub {
            my $direction = shift;
            return {}, {
               order_by => { "-$direction" => [qw{last_name first_name}] },
            };
         },
       }, sub {
      my $param = shift;
      my $direction = shift;
      return {}, {
         order_by => { "-$direction" => $param },
      };
       },$params
    );
 }

=head2 simple_deletion

 $self->simple_deletion({ rs => 'Foo' });

Deletes from the passed in resultset based on the following CGI parameter:

 to_delete - values of the ids of items to delete

Valid arguments are:

 rs - resultset loaded into schema

Note that this method uses the $rs->delete method, as opposed to $rs->delete_all

=head2 simple_search

 my $searched_rs = $self->simple_search({ rs => 'Foo' });

This method just searches on all of the CGI parameters that are not in the
C<ignored_params> with a like "%$value%".  If there are multiple values it will
make the search an C<or> between the different values.

Valid arguments are:

 rs - source loaded into schema

=head2 simple_sort

 my $resultset = $self->schema->resultset('Foo');
 my $sorted_rs = $self->simple_sort($resultset);

Sorts the passed in resultset based on the following CGI parameters:

 sort - field to sort by, defaults to primarky key
 dir  - direction to sort

=head1 SEE ALSO

L<CGI::Application::Plugin::DBH>

=head1 CREDITS

Thanks to Micro Technology Services, Inc. for funding the initial development
of this module.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__
